samples: bluetooth: mesh_provisioner: push button to provision

HaavardRei opened the pull request:
In order to keep a flashed sample from provisioning forever, the sample
requires the user to press a button to provision a detected device.

The feature is only enabled if the board has a push button connected
via GPIO.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

Andrewpini commented:
@jhedberg can you approve the workflow for this PR? This is a task from the Nordic Mesh team backlog that is done by one of our summer interns.

Andrewpini commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -294,6 +298,24 @@ static uint8_t check_unconfigured(struct bt_mesh_cdb_node *node, void *data)
+	       button.pin);
+}
+
+K_TIMER_DEFINE(timeout_timer, timeout_timer_expired, NULL);
```
Might sound a little bit strange, but in general we try to use `k_work_delayable` rather than timers. Could you try to duplicate the behaviour using `k_work_delayable`?

Andrewpini commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -311,33 +333,58 @@ void main(void)
 	printk("Bluetooth initialized\n");
 	bt_ready();
 
+	/* Initializing button to reset timeout */
+	int ret;
+
+	if (!device_is_ready(button.port)) {
+		printk("Error: button device %s is not ready\n", button.port->name);
+		return;
+	}
+	ret = gpio_pin_configure_dt(&button, GPIO_INPUT);
+	if (ret != 0) {
+		printk("Error %d: failed to configure %s pin %d\n", ret, button.port->name,
+		       button.pin);
+		return;
+	}
+	ret = gpio_pin_interrupt_configure_dt(&button, GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret != 0) {
+		printk("Error %d: failed to configure interrupt on %s pin %d\n", ret,
+		       button.port->name, button.pin);
+		return;
+	}
+	gpio_init_callback(&button_cb_data, button_pressed, BIT(button.pin));
+	gpio_add_callback(button.port, &button_cb_data);
+	printk("Press button at %s pin %d to enable Provisioner.\n", button.port->name, button.pin);
```
This could be wraped in a separate `button_init` function

Andrewpini replied:
```suggestion
	printk("Press button 1 to enable the Provisioner.\n);
```

Andrewpini reviewed (commented):
I have tested this and it seems to be working fine 👍 I have a couple of suggestions so please take a look at my comments.

Andrewpini reviewed (commented):
Looks good to me, but have you reviewed the README for the sample? Does it need to be updated after the added changes?

Andrewpini commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -294,6 +298,46 @@ static uint8_t check_unconfigured(struct bt_mesh_cdb_node *node, void *data)
+	k_work_reschedule(&timeout_work, K_MINUTES(3));
+}
+
+void button_init()
```
```suggestion
static void button_init(void)
```

Andrewpini commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -294,6 +298,46 @@ static uint8_t check_unconfigured(struct bt_mesh_cdb_node *node, void *data)
+	printk("Provisioner timed out, press button 1 to continue.\n");
+}
+
+void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
```
```suggestion
static void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
```

PavelVPV commented:
Wouldn't it be better to provision the found device only after button press?

Andrewpini commented:
> Wouldn't it be better to provision the found device only after button press?

Well, that would work to. I don't have a strong opinion on the approach as long as we get rid of the provision forever feature. If your up for it @HaavardRei you can try to do what @PavelVPV suggested.

HaavardRei commented:
@PavelVPV , @Andrewpini I reworked the code using a semaphore to synchronize provisioning with a button press. Should squash the previous commits before merging if this is approved.

alxelax commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -5,11 +5,15 @@
+
+#define SW0_NODE	DT_ALIAS(sw0)
+#if !DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#error "Unsupported board: sw0 devicetree alias is not defined"
```
What about platforms `**qemu_cortex_m3 qemu_x86**`? They are in the allowed platforms list.
Is it still possible to build and run app for them?

alxelax replied:
I'm not sure about the importance of platforms. This is a question mostly to the code owner.
However, I think if it is not possible to build an application for the board it shouldn't be in the list of allowed platforms.

PavelVPV replied:
(not a code owner of this sample) No, it won't build. But a user may want to run this sample in QEMU or as native application and use BlueZ as a controller. This variant is also mentioned in the documentation. As a suggestion, it can either be compiled with button option for boards that has a button and without button option for other boards (like QEMU).

HaavardRei replied:
I've done as @PavelVPV suggested and in the latest commit the feature is only enabled if the board supports GPIO. The sample builds for all specified platforms.

PavelVPV commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -348,7 +349,9 @@ void main(void)
 	printk("Bluetooth initialized\n");
 	bt_ready();
 
+	#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
```
Minor: If I'm not mistaken, preprocessor directives should not have indentation.

PavelVPV commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -363,13 +366,15 @@ void main(void)
+		#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
 		k_sem_reset(&sem_button_pressed);
 		printk("Device %s detected, press button 1 to provision.\n", uuid_hex_str);
 		err = k_sem_take(&sem_button_pressed, K_MINUTES(3));
```
Minor: 3 minutes can be too much. May be 30 seconds?

PavelVPV reviewed (changes requested):
Few last minor comments from my side.

You can now update the PR description and title according to the current changes.
Also, feel free to force push commits, github will show delta to the previous changes.

HaavardRei committed:
samples: bluetooth: mesh_provisioner: push button to provision

In order to keep a flashed sample from provisioning forever, the sample
requires the user to press a button to provision a detected device.

The feature is only enabled if the board has a push button connected
via GPIO.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

Andrewpini reviewed (approved):
LGTM 👍

Andrewpini commented:
I see that compliance is complaining about a couple of long lines in the documentation. Please take a look and resolve it.

PavelVPV commented:
@HaavardRei you need to shorten lines in commit message to 75 symbols.

HaavardRei commented:
@PavelVPV , @alxelax It should be good now :)
